File: studio/studio-domain/src/commonMain/kotlin/com/moqserver/studio/domain/ImportConverter.kt
package com.moqserver.studio.domain

import com.moqserver.studio.projectformat.AuthType
import com.moqserver.studio.projectformat.EndpointDocument
import com.moqserver.studio.projectformat.MoqProject
import com.moqserver.studio.projectformat.NetworkBehavior
import com.moqserver.studio.projectformat.ProjectAuthConfig
import com.moqserver.studio.projectformat.ProjectDefaults
import com.moqserver.studio.projectformat.ProjectManifest
import com.moqserver.studio.projectformat.ProjectVariant
import com.moqserver.studio.projectformat.RequestRules
import com.moqserver.studio.projectformat.RuleMatcher
import com.moqserver.studio.projectformat.YamlValue
import com.moqserver.studio.projectformat.defaultAliasForEndpoint
import com.moqserver.studio.projectformat.suggestedEndpointReferenceName
import com.moqserver.studio.projectformat.suggestedVariantName
import com.moqserver.studio.projectformat.suggestedVariantReferenceName
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.longOrNull

/** Converts a ParsedSpec into a MoqProject, filtering to accepted endpoints only. */
object ImportConverter {
    private val json = Json { ignoreUnknownKeys = true }

    fun convert(
        spec: ParsedSpec,
        acceptedEntries: List<ImportEndpointEntry>,
        projectName: String,
        projectPath: String,
    ): MoqProject {
        val assignedEndpointReferenceNames = mutableListOf<String>()
        val endpoints = acceptedEntries.map { convertEndpoint(it, assignedEndpointReferenceNames) }
        val manifest = ProjectManifest(
            version = "1",
            name = projectName,
            description = "Imported from ${spec.title} (v${spec.version})",
            defaults = ProjectDefaults(
                delayMs = 0,
                auth = ProjectAuthConfig(type = AuthType.NONE, verify = false),
                network = NetworkBehavior(),
            ),
        )
        return MoqProject(
            manifest = manifest,
            endpoints = endpoints,
            projectPath = projectPath,
        )
    }

    /**
     * Merges accepted import entries into an existing project.
     *
     * Strategy:
     * - NEW entries: converted fresh and appended to the project endpoint list.
     * - CHANGED entries: spec-owned fields (auth, request rules, tags) are updated on the existing
     *   endpoint; only new response status codes from the spec are added as new variants — existing
     *   user-authored variants are preserved untouched.
     * - UNCHANGED entries: always skipped (they contribute nothing new to the project).
     * - Entries with [accepted=false] are always skipped regardless of status.
     */
    fun merge(
        existingProject: MoqProject,
        acceptedEntries: List<ImportEndpointEntry>,
        updateSelection: UpdateSelection = UpdateSelection(),
    ): MoqProject {
        val newEntries = acceptedEntries.filter {
            it.updateStatus == EndpointUpdateStatus.NEW && updateSelection.url
        }
        val changedEntries = acceptedEntries.filter {
            it.updateStatus == EndpointUpdateStatus.CHANGED && shouldMergeChangedEntry(it, updateSelection)
        }

        // Build set of existing endpoint IDs for collision avoidance when assigning new reference names
        val assignedEndpointReferenceNames = existingProject.endpoints.map { it.referenceName }.toMutableList()

        // Convert brand-new endpoints fresh
        val convertedNew = newEntries.map { convertEndpoint(it, assignedEndpointReferenceNames) }

        // Apply spec-field updates to changed endpoints (preserving user content)
        val updatedExisting = existingProject.endpoints.map { existing ->
            val changedEntry = changedEntries.find { entry ->
                endpointId(entry.endpoint.method, entry.endpoint.path) == existing.id
            }
            if (changedEntry != null) mergeEndpoint(existing, changedEntry, updateSelection) else existing
        }

        return existingProject.copy(
            endpoints = updatedExisting + convertedNew,
        )
    }

    /**
     * Computes a [EndpointSpecDiff] by comparing a freshly parsed endpoint against an existing
     * [EndpointDocument]. Only spec-owned, non-user-editable fields are compared.
     */
    fun diffEndpoint(parsed: ParsedEndpoint, existing: EndpointDocument): EndpointSpecDiff {
        val parsedStatusCodes = parsed.responses.map { it.statusCode }.toSet()
        val existingStatusCodes = existing.variants.map { it.status }.toSet()
        val newStatusCodes = parsedStatusCodes - existingStatusCodes
        val removedStatusCodes = existingStatusCodes - parsedStatusCodes

        val parsedAuth = if (parsed.authType != AuthType.NONE) {
            parsed.authType to parsed.authHeaderName
        } else {
            null
        }
        val existingAuth = existing.auth?.let { it.type to it.headerName }
        val authChanged = parsedAuth != existingAuth

        val parsedRequiredHeaders = parsed.requiredHeaders.toSet()
        val existingRequiredHeaders = existing.requestRules?.headers
            ?.filter { it.required == true }
            ?.map { it.name }
            ?.toSet()
            ?: emptySet()

        val parsedQueryParams = if (parsed.queryParameters.isNotEmpty()) {
            parsed.queryParameters.map { it.name }.toSet()
        } else {
            parsed.requiredQueryParameters.toSet()
        }
        val existingQueryParams = existing.requestRules?.queryParams
            ?.map { it.name }
            ?.toSet()
            ?: emptySet()

        val parsedCookies = parsed.cookies.map { it.name }.toSet()
        val existingCookies = existing.requestRules?.cookies
            ?.map { it.name }
            ?.toSet()
            ?: emptySet()

        val requestRulesChanged = parsedRequiredHeaders != existingRequiredHeaders ||
            parsedQueryParams != existingQueryParams ||
            parsedCookies != existingCookies

        val parsedTags = parsed.tags.toSet()
        val existingTags = existing.tags?.toSet() ?: emptySet()
        val tagsChanged = parsedTags != existingTags

		val parsedBodiesByStatus = parsed.responses.associate { response ->
			response.statusCode to normalizeParsedBody(response.body)
		}
		val existingBodiesByStatus = existing.variants.associate {
			it.status to normalizeBody(it.body?.let(::yamlBodyToComparableString))
		}
        val sharedStatusCodes = parsedBodiesByStatus.keys intersect existingBodiesByStatus.keys
        val responseBodyChanged = sharedStatusCodes.any { status ->
            parsedBodiesByStatus[status] != existingBodiesByStatus[status]
        }

		return EndpointSpecDiff(
			newStatusCodes = newStatusCodes,
			removedStatusCodes = removedStatusCodes,
			authChanged = authChanged,
			requestRulesChanged = requestRulesChanged,
			tagsChanged = tagsChanged,
			responseBodyChanged = responseBodyChanged,
		)
	}

    /**
     * Merges spec-owned fields from a [changedEntry] into an [existing] endpoint.
     *
     * Preserved from existing (user-owned): alias, description, referenceName, body/bodyFile,
     * delayMs, requestMatch, isDefault, variant names/reference names.
     *
     * Updated from spec (spec-owned): auth, requestRules, tags.
     * Added from spec: variants whose status codes are new in the spec (existing variants kept).
     */
    @Suppress("LongMethod")
    private fun mergeEndpoint(
        existing: EndpointDocument,
        changedEntry: ImportEndpointEntry,
        updateSelection: UpdateSelection,
    ): EndpointDocument {
        val parsed = changedEntry.endpoint
        val diff = changedEntry.specDiff

        // Update spec-owned structural fields
        val updatedAuth = if (parsed.authType != AuthType.NONE) {
            ProjectAuthConfig(
                type = parsed.authType,
                verify = true,
                headerName = parsed.authHeaderName,
            )
        } else {
            null
        }

        val updatedRequestRules = if (updateSelection.details && diff?.requestRulesChanged == true) {
            buildRequestRules(parsed)
        } else {
            existing.requestRules
        }

        val updatedTags = if (updateSelection.details && diff?.tagsChanged == true) {
            parsed.tags.ifEmpty { null }
        } else {
            existing.tags
        }

        // Add or replace variants when body updates are enabled; otherwise preserve existing variants.
        val existingStatusCodes = existing.variants.map { it.status }.toSet()
        val mergedVariants = if (updateSelection.body) {
            mergeVariants(existing, parsed, changedEntry.generatedResponses)
        } else {
            existing.variants
        }

		return existing.copy(
			auth = if (updateSelection.details && diff?.authChanged == true) updatedAuth else existing.auth,
			requestRules = updatedRequestRules,
			tags = updatedTags,
			variants = mergedVariants,
		)
	}

	private fun shouldMergeChangedEntry(entry: ImportEndpointEntry, updateSelection: UpdateSelection): Boolean {
		val diff = entry.specDiff ?: return false
		return (updateSelection.details && diff.affectsDetails) ||
			(updateSelection.body && (diff.affectsBody || entry.generatedResponses.isNotEmpty()))
	}

	private fun mergeVariants(
		existing: EndpointDocument,
		parsed: ParsedEndpoint,
		generatedResponses: List<ParsedResponse>,
	): List<ProjectVariant> {
		val responses = deduplicateResponses(parsed.responses + generatedResponses)
		val responsesByStatus = responses.groupBy { it.statusCode }.mapValues { (_, values) -> values.toMutableList() }

		val assignedNames = existing.variants.map { it.name }.toMutableList()
		val assignedRefNames = existing.variants.map { it.referenceName }.toMutableList()

		val updatedExisting = existing.variants.map { variant ->
			val candidates = responsesByStatus[variant.status] ?: return@map variant
			val replacement = candidates.firstOrNull { it.name == variant.name }
				?: candidates.singleOrNull()
				?: return@map variant
			candidates.remove(replacement)
			convertVariant(
				resp = replacement,
				name = variant.name,
				referenceName = variant.referenceName,
				isDefault = variant.isDefault == true,
			).copy(
				requestMatch = variant.requestMatch,
				delayMs = variant.delayMs,
				bodyFile = variant.bodyFile,
			)
		}

		val newResponses = responsesByStatus.values.flatten()
		val newVariants = newResponses.map { resp ->
			val name = suggestedVariantName(
				status = resp.statusCode,
				existingNames = assignedNames,
				preferredName = resp.name,
			)
			assignedNames += name
			val refName = suggestedVariantReferenceName(
				preferredSource = name,
				status = resp.statusCode,
				existingNames = assignedRefNames,
			)
			assignedRefNames += refName
			convertVariant(resp = resp, name = name, referenceName = refName, isDefault = false)
		}

		return updatedExisting + newVariants
	}

	private fun normalizeBody(body: String?): String? = body?.trim()?.ifBlank { null }

	private fun normalizeParsedBody(body: String?): String? {
		if (body == null) return null
		return parseBodyToYamlValue(body).let(::yamlBodyToComparableString).trim().ifBlank { null }
	}

	private fun yamlBodyToComparableString(body: YamlValue): String = when (body) {
		is YamlValue.Null -> "null"
		is YamlValue.Str -> body.value
		is YamlValue.Int -> body.value.toString()
		is YamlValue.Double -> body.value.toString()
		is YamlValue.Bool -> body.value.toString()
		is YamlValue.Array ->
			body.value.joinToString(prefix = "[", postfix = "]") { yamlBodyToComparableString(it) }
		is YamlValue.Obj ->
			body.value.entries
				.sortedBy { it.key }
				.joinToString(prefix = "{", postfix = "}") { (key, value) ->
					"$key:${yamlBodyToComparableString(value)}"
				}
	}

	@Suppress("LongMethod")
	private fun convertEndpoint(
        entry: ImportEndpointEntry,
        assignedEndpointReferenceNames: MutableList<String>,
    ): EndpointDocument {
        val parsed = entry.endpoint
        val id = endpointId(parsed.method, parsed.path)
        val alias = parsed.alias?.takeIf { it.isNotBlank() }
            ?: defaultAliasForEndpoint(method = parsed.method, path = parsed.path)
        val referenceName = suggestedEndpointReferenceName(
            preferredSource = parsed.referenceName ?: alias,
            fallbackId = id,
            existingNames = assignedEndpointReferenceNames,
        )
        assignedEndpointReferenceNames += referenceName
		val allResponses = deduplicateResponses(parsed.responses + entry.generatedResponses)
        val defaultIndex = when {
            allResponses.isEmpty() -> -1
            else -> allResponses.indexOfFirst { it.name == "default" }.takeIf { it >= 0 }
                ?: allResponses.indexOfFirst { it.statusCode in 200..299 }.takeIf { it >= 0 }
                ?: 0
        }
        val assignedNames = mutableListOf<String>()
        val assignedReferenceNames = mutableListOf<String>()
        val variants = allResponses.mapIndexed { index, resp ->
            val name = suggestedVariantName(
                status = resp.statusCode,
                existingNames = assignedNames,
                preferredName = resp.name,
            )
            assignedNames += name
            val variantReferenceName = suggestedVariantReferenceName(
                preferredSource = name,
                status = resp.statusCode,
                existingNames = assignedReferenceNames,
            )
            assignedReferenceNames += variantReferenceName
            convertVariant(
                resp = resp,
                name = name,
                referenceName = variantReferenceName,
                isDefault = index == defaultIndex,
            )
        }

        val auth = if (parsed.authType != AuthType.NONE) {
            ProjectAuthConfig(
                type = parsed.authType,
                verify = true,
                headerName = parsed.authHeaderName,
            )
        } else {
            null
        }

        val requestRules = buildRequestRules(parsed)

        return EndpointDocument(
            id = id,
            alias = alias,
            description = parsed.description?.takeIf { it.isNotBlank() },
            referenceName = referenceName,
            method = parsed.method.uppercase(),
            path = parsed.path,
            tags = parsed.tags.ifEmpty { null },
            auth = auth,
            requestRules = requestRules,
            variants = variants,
        )
    }

    private fun convertVariant(
        resp: ParsedResponse,
        name: String,
        referenceName: String,
        isDefault: Boolean,
    ): ProjectVariant {
        val body = resp.body?.let { parseBodyToYamlValue(it) }
        return ProjectVariant(
            name = name,
            referenceName = referenceName,
            description = resp.description,
            isDefault = if (isDefault) true else null,
            status = resp.statusCode,
            headers = resp.headers.ifEmpty { null },
            body = body,
        )
    }

	private fun parseBodyToYamlValue(body: String): YamlValue {
        val trimmed = body.trim()
        if (trimmed.isEmpty()) return YamlValue.Str(body)

        return runCatching { json.parseToJsonElement(trimmed) }
            .map { jsonElementToYamlValue(it) }
            .getOrElse { YamlValue.Str(body) }
	}

	private fun deduplicateResponses(responses: List<ParsedResponse>): List<ParsedResponse> {
		return responses
			.groupBy { response -> response.statusCode to normalizeParsedBody(response.body) }
			.values
			.map { duplicates -> duplicates.last() }
	}

    private fun jsonElementToYamlValue(element: JsonElement): YamlValue {
        return when (element) {
            is JsonNull -> YamlValue.Null
            is JsonPrimitive -> {
                when {
                    element.isString -> YamlValue.Str(element.content)
                    element.booleanOrNull != null -> YamlValue.Bool(element.booleanOrNull!!)
                    element.intOrNull != null -> YamlValue.Int(element.intOrNull!!)
                    element.longOrNull != null -> YamlValue.Double(element.longOrNull!!.toDouble())
                    element.doubleOrNull != null -> YamlValue.Double(element.doubleOrNull!!)
                    else -> YamlValue.Str(element.content)
                }
            }
            is JsonArray ->
                YamlValue.Array(element.map { jsonElementToYamlValue(it) })
            is JsonObject ->
                YamlValue.Obj(element.entries.associate { (k, v) -> k to jsonElementToYamlValue(v) })
        }
    }

    private fun buildRequestRules(parsed: ParsedEndpoint): RequestRules? {
        val headers = parsed.requiredHeaders.map { RuleMatcher(name = it, required = true) }
            .ifEmpty { null }
        val queryParams = parsed.queryParameters
            .ifEmpty {
                parsed.requiredQueryParameters.map { RuleMatcher(name = it, required = true) }
            }
            .ifEmpty { null }
        val cookies = parsed.cookies.ifEmpty { null }

        return if (headers != null || queryParams != null || cookies != null) {
            RequestRules(headers = headers, queryParams = queryParams, cookies = cookies)
        } else {
            null
        }
    }

    /** Generate a deterministic endpoint ID from method + path. */
    fun endpointId(method: String, path: String): String {
        val normalized = path
            .removePrefix("/")
            .replace(Regex("\\{[^}]+}"), "param")
            .replace(Regex("[^a-zA-Z0-9/]"), "")
            .replace("/", "-")
            .lowercase()
            .trimEnd('-')

        val id = "${method.lowercase()}-$normalized"
        return id.ifEmpty { "${method.lowercase()}-root" }
    }
}

File: server/Sources/MoqImport/ImportConverter.swift
import Foundation
import MoqCore

/// Which parsed endpoints participate in a conversion or merge, computed by the caller ahead of
/// time (e.g. from user-selected paths/tags, or "accept everything"). Filters are intersected:
/// an endpoint must match every supplied filter; an omitted filter accepts all.
public struct ImportSelection: Sendable {
    public let acceptedPaths: Set<String>?
    public let acceptedTags: Set<String>?

    public init(acceptedPaths: Set<String>? = nil, acceptedTags: Set<String>? = nil) {
        self.acceptedPaths = acceptedPaths
        self.acceptedTags = acceptedTags
    }

    public static let all = ImportSelection()

    func accepts(_ endpoint: ParsedEndpoint) -> Bool {
        if let acceptedPaths, !acceptedPaths.contains(endpoint.path) { return false }
        if let acceptedTags, Set(endpoint.tags).isDisjoint(with: acceptedTags) { return false }
        return true
    }
}

/// Governs how `ImportConverter.merge` treats endpoints that already exist in the target
/// project.
public struct ImportMergePolicy: Sendable {
    /// Update spec-owned fields (auth, request rules, tags) on endpoints that already exist.
    /// User-owned fields (alias, description, reference_name, defaults, delay, request_match,
    /// isDefault, variant names/reference names) are always preserved regardless of this flag.
    public let updateDetails: Bool
    /// Overwrite the body of an existing variant when the spec has a response at the same
    /// status code with a different body. When `false` (the default), existing variant bodies
    /// are never touched — only genuinely new status-code variants are added.
    public let replaceExistingBodies: Bool

    public init(updateDetails: Bool = true, replaceExistingBodies: Bool = false) {
        self.updateDetails = updateDetails
        self.replaceExistingBodies = replaceExistingBodies
    }
}

/// Converts a `ParsedSpec` into a `MoqProject`, or merges it into an existing one.
///
/// This intentionally simplifies Kotlin's `ImportConverter`
/// (`studio/studio-domain/src/commonMain/kotlin/com/moqserver/studio/domain/ImportConverter.kt`):
/// there is no interactive per-endpoint review step here (no `ImportEndpointEntry.accepted`, no
/// AI-generated response selection — that stays a Studio-only workflow). `ImportSelection`
/// replaces per-entry acceptance, and endpoint NEW/CHANGED/UNCHANGED classification happens
/// internally in `merge` by diffing against the target project, rather than being precomputed
/// for a review UI to display.
public enum ImportConverter {
    public static func convert(_ spec: ParsedSpec, selection: ImportSelection, name: String, path: String) -> MoqProject {
        var assignedReferenceNames: [String] = []
        let accepted = spec.endpoints.filter(selection.accepts)
        let endpoints = accepted.map { convertEndpoint($0, assignedReferenceNames: &assignedReferenceNames) }

        let manifest = ProjectManifest(
            version: MoqFormatRules.formatVersion,
            name: name,
            description: "Imported from \(spec.title) (v\(spec.version))",
            defaults: ProjectDefaults(
                delayMs: 0,
                auth: ProjectAuthConfig(type: .none, verify: false),
                network: NetworkBehavior()
            )
        )
        return MoqProject(manifest: manifest, endpoints: endpoints, projectPath: path)
    }

    /// Merges accepted import entries into an existing project.
    ///
    /// - New endpoints (no existing endpoint at the same id) are converted fresh and appended.
    /// - Existing endpoints never lose variants or fields merely because the spec omits or
    ///   changes them: spec-owned fields (auth, request rules, tags) are only updated when
    ///   `policy.updateDetails` is true, and existing variant bodies are only overwritten when
    ///   `policy.replaceExistingBodies` is true. New status-code variants from the spec are
    ///   always added.
    public static func merge(
        _ spec: ParsedSpec, selection: ImportSelection, policy: ImportMergePolicy, into existing: MoqProject
    ) -> MoqProject {
        let accepted = spec.endpoints.filter(selection.accepts)
        var existingByID: [String: EndpointDocument] = [:]
        for endpoint in existing.endpoints { existingByID[endpoint.id] = endpoint }

        var assignedReferenceNames = existing.endpoints.map(\.referenceName)
        var mergedEndpoints = existing.endpoints
        var newEndpoints: [EndpointDocument] = []

        for parsed in accepted {
            let id = endpointID(method: parsed.method, path: parsed.path)
            if let index = mergedEndpoints.firstIndex(where: { $0.id == id }) {
                mergedEndpoints[index] = mergeEndpoint(mergedEndpoints[index], parsed: parsed, policy: policy)
            } else {
                newEndpoints.append(convertEndpoint(parsed, assignedReferenceNames: &assignedReferenceNames))
            }
        }

        return MoqProject(
            manifest: existing.manifest, endpoints: mergedEndpoints + newEndpoints, projectPath: existing.projectPath)
    }

    /// Compares a freshly parsed endpoint against an existing one, reporting only spec-owned,
    /// non-user-editable differences.
    public static func diff(_ parsed: ParsedEndpoint, existing: EndpointDocument) -> EndpointSpecDiff {
        let parsedStatusCodes = Set(parsed.responses.map(\.statusCode))
        let existingStatusCodes = Set(existing.variants.map(\.status))
        let newStatusCodes = parsedStatusCodes.subtracting(existingStatusCodes)
        let removedStatusCodes = existingStatusCodes.subtracting(parsedStatusCodes)

        let parsedAuth: (ProjectAuthConfig.AuthType, String?)? =
            parsed.authType != .none ? (parsed.authType, parsed.authHeaderName) : nil
        let existingAuth: (ProjectAuthConfig.AuthType, String?)? = existing.auth.map { ($0.type, $0.headerName) }
        let authChanged = !authPairsEqual(parsedAuth, existingAuth)

        let parsedRequiredHeaders = Set(parsed.requiredHeaders)
        let existingRequiredHeaders = Set(
            (existing.requestRules?.headers ?? []).filter { $0.required == true }.map(\.name))

        let parsedQueryParams =
            !parsed.queryParameters.isEmpty
            ? Set(parsed.queryParameters.map(\.name)) : Set(parsed.requiredQueryParameters)
        let existingQueryParams = Set((existing.requestRules?.queryParams ?? []).map(\.name))

        let parsedCookies = Set(parsed.cookies.map(\.name))
        let existingCookies = Set((existing.requestRules?.cookies ?? []).map(\.name))

        let requestRulesChanged =
            parsedRequiredHeaders != existingRequiredHeaders || parsedQueryParams != existingQueryParams
            || parsedCookies != existingCookies

        let tagsChanged = Set(parsed.tags) != Set(existing.tags ?? [])

        return EndpointSpecDiff(
            newStatusCodes: newStatusCodes,
            removedStatusCodes: removedStatusCodes,
            authChanged: authChanged,
            requestRulesChanged: requestRulesChanged,
            tagsChanged: tagsChanged
        )
    }

    private static func authPairsEqual(
        _ lhs: (ProjectAuthConfig.AuthType, String?)?, _ rhs: (ProjectAuthConfig.AuthType, String?)?
    ) -> Bool {
        switch (lhs, rhs) {
        case (nil, nil): return true
        case (let l?, let r?): return l.0 == r.0 && l.1 == r.1
        default: return false
        }
    }

    // MARK: - Merge

    private static func mergeEndpoint(
        _ existing: EndpointDocument, parsed: ParsedEndpoint, policy: ImportMergePolicy
    ) -> EndpointDocument {
        let diff = diff(parsed, existing: existing)

        let auth: ProjectAuthConfig? = (policy.updateDetails && diff.authChanged) ? parsed.toProjectAuthConfig() : existing.auth
        let requestRules: RequestRules? =
            (policy.updateDetails && diff.requestRulesChanged) ? buildRequestRules(parsed) : existing.requestRules
        let tags: [String]? =
            (policy.updateDetails && diff.tagsChanged) ? (parsed.tags.isEmpty ? nil : parsed.tags) : existing.tags

        return EndpointDocument(
            id: existing.id,
            alias: existing.alias,
            description: existing.description,
            referenceName: existing.referenceName,
            method: existing.method,
            path: existing.path,
            tags: tags,
            auth: auth,
            requestRules: requestRules,
            operation: existing.operation,
            network: existing.network,
            variants: mergeVariants(existing, parsed: parsed, policy: policy)
        )
    }

    /// Existing variants are always kept, and a status code already covered by an existing
    /// variant never spawns a duplicate "new" variant for that status — `replaceExistingBodies`
    /// only controls whether that existing variant's content gets overwritten in place. Status
    /// codes with no existing variant are always appended as new variants.
    private static func mergeVariants(
        _ existing: EndpointDocument, parsed: ParsedEndpoint, policy: ImportMergePolicy
    ) -> [ProjectVariant] {
        let existingStatusCodes = Set(existing.variants.map(\.status))

        var responsesByStatus: [Int: [ParsedResponse]] = [:]
        var statusOrder: [Int] = []
        for response in parsed.responses {
            if responsesByStatus[response.statusCode] == nil { statusOrder.append(response.statusCode) }
            responsesByStatus[response.statusCode, default: []].append(response)
        }

        var updatedExisting = existing.variants
        if policy.replaceExistingBodies {
            updatedExisting = existing.variants.map { variant -> ProjectVariant in
                guard var candidates = responsesByStatus[variant.status], !candidates.isEmpty else { return variant }
                let replacementIndex = candidates.firstIndex { $0.name == variant.name } ?? (candidates.count == 1 ? 0 : nil)
                guard let replacementIndex else { return variant }
                let replacement = candidates.remove(at: replacementIndex)
                responsesByStatus[variant.status] = candidates
                let replacementBody = replacement.body.map(parseBody)
                return ProjectVariant(
                    name: variant.name,
                    referenceName: variant.referenceName,
                    description: replacement.description ?? variant.description,
                    isDefault: variant.isDefault,
                    status: variant.status,
                    headers: replacement.headers.isEmpty ? variant.headers : replacement.headers,
                    requestMatch: variant.requestMatch,
                    body: replacementBody,
                    bodyFile: replacementBody != nil ? nil : variant.bodyFile,
                    delayMs: variant.delayMs
                )
            }
        }

        var assignedNames = existing.variants.map(\.name)
        var assignedReferenceNames = existing.variants.map(\.referenceName)
        let newResponses = statusOrder.filter { !existingStatusCodes.contains($0) }.flatMap { responsesByStatus[$0] ?? [] }
        let newVariants = newResponses.map { response -> ProjectVariant in
            let name = suggestedVariantName(status: response.statusCode, existingNames: assignedNames, preferredName: response.name)
            assignedNames.append(name)
            let referenceName = suggestedVariantReferenceName(
                preferredSource: name, status: response.statusCode, existingNames: assignedReferenceNames)
            assignedReferenceNames.append(referenceName)
            return convertVariant(response, name: name, referenceName: referenceName, isDefault: false)
        }

        return updatedExisting + newVariants
    }

    // MARK: - Fresh conversion

    private static func convertEndpoint(_ parsed: ParsedEndpoint, assignedReferenceNames: inout [String]) -> EndpointDocument {
        let id = endpointID(method: parsed.method, path: parsed.path)
        let alias = parsed.alias?.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty
            ?? EndpointAlias.defaultAlias(method: parsed.method, path: parsed.path)
        let referenceName = suggestedEndpointReferenceName(
            preferredSource: parsed.referenceName ?? alias, fallbackID: id, existingNames: assignedReferenceNames)
        assignedReferenceNames.append(referenceName)

        var variantNames: [String] = []
        var variantReferenceNames: [String] = []
        let defaultIndex = defaultVariantIndex(parsed.responses)
        let variants = parsed.responses.enumerated().map { index, response -> ProjectVariant in
            let name = suggestedVariantName(status: response.statusCode, existingNames: variantNames, preferredName: response.name)
            variantNames.append(name)
            let refName = suggestedVariantReferenceName(
                preferredSource: name, status: response.statusCode, existingNames: variantReferenceNames)
            variantReferenceNames.append(refName)
            return convertVariant(response, name: name, referenceName: refName, isDefault: index == defaultIndex)
        }

        return EndpointDocument(
            id: id,
            alias: alias,
            description: parsed.description?.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty,
            referenceName: referenceName,
            method: parsed.method.uppercased(),
            path: parsed.path,
            tags: parsed.tags.isEmpty ? nil : parsed.tags,
            auth: parsed.toProjectAuthConfig(),
            requestRules: buildRequestRules(parsed),
            variants: variants
        )
    }

    private static func defaultVariantIndex(_ responses: [ParsedResponse]) -> Int {
        guard !responses.isEmpty else { return -1 }
        if let index = responses.firstIndex(where: { $0.name == "default" }) { return index }
        if let index = responses.firstIndex(where: { (200...299).contains($0.statusCode) }) { return index }
        return 0
    }

    private static func convertVariant(
        _ response: ParsedResponse, name: String, referenceName: String, isDefault: Bool
    ) -> ProjectVariant {
        ProjectVariant(
            name: name,
            referenceName: referenceName,
            description: response.description,
            isDefault: isDefault ? true : nil,
            status: response.statusCode,
            headers: response.headers.isEmpty ? nil : response.headers,
            body: response.body.map(parseBody)
        )
    }

    private static func buildRequestRules(_ parsed: ParsedEndpoint) -> RequestRules? {
        let headers = parsed.requiredHeaders.map { RuleMatcher(name: $0, required: true) }
        let queryParams =
            !parsed.queryParameters.isEmpty
            ? parsed.queryParameters : parsed.requiredQueryParameters.map { RuleMatcher(name: $0, required: true) }
        let cookies = parsed.cookies

        guard !headers.isEmpty || !queryParams.isEmpty || !cookies.isEmpty else { return nil }
        return RequestRules(
            headers: headers.isEmpty ? nil : headers,
            queryParams: queryParams.isEmpty ? nil : queryParams,
            cookies: cookies.isEmpty ? nil : cookies
        )
    }

    /// Generates a deterministic endpoint id from method + path.
    public static func endpointID(method: String, path: String) -> String {
        var normalized = path
        if normalized.hasPrefix("/") { normalized.removeFirst() }
        normalized = normalized.replacingOccurrences(of: #"\{[^}]+\}"#, with: "param", options: .regularExpression)
        normalized = normalized.replacingOccurrences(of: "[^a-zA-Z0-9/]", with: "", options: .regularExpression)
        normalized = normalized.replacingOccurrences(of: "/", with: "-")
        normalized = normalized.lowercased()
        while normalized.hasSuffix("-") { normalized.removeLast() }

        let id = "\(method.lowercased())-\(normalized)"
        return normalized.isEmpty ? "\(method.lowercased())-root" : id
    }

    /// Parses a raw response body string into a structured value: JSON text becomes a decoded
    /// `AnyCodableValue` tree; anything that fails to parse as JSON (including plain text) is
    /// kept as a `.string`.
    private static func parseBody(_ raw: String) -> AnyCodableValue {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return .string(raw) }
        guard let data = trimmed.data(using: .utf8),
            let object = try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
        else {
            return .string(raw)
        }
        return jsonObjectToAnyCodableValue(object)
    }

    private static func jsonObjectToAnyCodableValue(_ object: Any) -> AnyCodableValue {
        switch object {
        case is NSNull:
            return .null
        case let number as NSNumber:
            if CFGetTypeID(number) == CFBooleanGetTypeID() {
                return .bool(number.boolValue)
            }
            if number.stringValue.contains(".") || number.stringValue.lowercased().contains("e") {
                return .double(number.doubleValue)
            }
            return .int(number.intValue)
        case let string as String:
            return .string(string)
        case let array as [Any]:
            return .array(array.map(jsonObjectToAnyCodableValue))
        case let dict as [String: Any]:
            return .object(dict.mapValues(jsonObjectToAnyCodableValue))
        default:
            return .null
        }
    }
}

extension ParsedEndpoint {
    fileprivate func toProjectAuthConfig() -> ProjectAuthConfig? {
        guard authType != .none else { return nil }
        return ProjectAuthConfig(type: authType, verify: true, headerName: authHeaderName)
    }
}

extension String {
    fileprivate var nonEmpty: String? { isEmpty ? nil : self }
}

/// Describes which spec-owned changes were detected between a freshly parsed endpoint and an
/// existing project endpoint.
public struct EndpointSpecDiff: Sendable, Equatable {
    public let newStatusCodes: Set<Int>
    public let removedStatusCodes: Set<Int>
    public let authChanged: Bool
    public let requestRulesChanged: Bool
    public let tagsChanged: Bool

    public init(
        newStatusCodes: Set<Int> = [], removedStatusCodes: Set<Int> = [], authChanged: Bool = false,
        requestRulesChanged: Bool = false, tagsChanged: Bool = false
    ) {
        self.newStatusCodes = newStatusCodes
        self.removedStatusCodes = removedStatusCodes
        self.authChanged = authChanged
        self.requestRulesChanged = requestRulesChanged
        self.tagsChanged = tagsChanged
    }

    public var hasChanges: Bool {
        !newStatusCodes.isEmpty || !removedStatusCodes.isEmpty || authChanged || requestRulesChanged || tagsChanged
    }
}
